sklearn/kernel_approximation.py

```diff
@@ -360,6 +363,12 @@ def fit(self, X, y=None):
         )
 
         self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
```

Comment:
E.g.

```suggestion
        self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components).astype(
            X.dtype, copy=False
        )
```

Reply:
I do not think we can do this because `X.dtype` can be an `int` here. `_validate_data` does not cast `X` into a float array.

Reply:
Not if you force the casting in validate data. Does it make sense do apply an rbf sampler on integer data?

Reply:
I'm mostly concern with the memory regression: https://github.com/scikit-learn/scikit-learn/pull/24317#discussion_r963045400

> Does it make sense do apply an rbf sampler on integer data?

I think it makes sense for any numerical data. The input dtype can be int, but `transform` always returns a float array.

Reply:
For this estimator, since fit does not use X for any computation, I think it makes sense to keep the default dtype in validate_data, i.e. ``dtype="numeric"``, not converting any numeric dtype. So we should only convert do X's dtype when it's float32, so the current version is good

Reply:
I was also thinking that you only want to use `float32` for `random_offset_` and `random_weights_` if `X` is of that type. And for all other types of `X` you keep the current behaviour.

Reply:
> I was also thinking that you only want to use float32 for random_offset_ and random_weights_ if X is of that type. And for all other types of X you keep the current behaviour.

I agree on the point where we should restrain to `np.float64` and `np.float32` and keep the `dtype="numeric"`.

Regarding the default dtype for lower precision than 32-bits, I am wondering if this is not counter-productive to output a 64 bits array (and store 64-bits attributes). If we head to be memory efficient, then it would make more sense to use 32 bits as a default for these lower precision dtype, isn't it?

I don't recall how much consistent we are in other estimators. This is something that we should look at before enforcing it.

Reply:
Can we address "32bit to save memory/speed up compute when the input isn't float64" in a different PR? I think it is a good thing to think about but it could also lead to changes in the user facing output or some other "can of worms". So maybe something for a different PR/concerted effort across all of scikit-learn?

Reply:
> Can we address "32bit to save memory/speed up compute when the input isn't float64" in a different PR?

I agree, let's think about that in another PR and keep the focus on just preserving float32